invenio_rdm_records/services/pids/providers/base.py

```diff
@@ -164,3 +164,17 @@ def validate(self, record, identifier=None, provider=None, **kwargs):
             pass
 
         return True, []
+
+    def validate_record(self, record):
```

Comment:
As discussed in the previous PR, IMO this creates more confusion than clarity.
The `validate` method above is enough to validate PID, record and anything else. It was designed exactly for these use cases.
@zzacharo shared the same concerns about duplicating this method.

It would be nice to have a comment from other reviewers too (@tmorrell maybe, given that you have worked extensively around this).

Reply:
I am so sorry, but I think there was a misunderstanding. I understood [the discussion](https://github.com/inveniosoftware/invenio-rdm-records/pull/1140/files#diff-82789e81eafd9ec8071709d9a32ecd6b52c83cb1b9bf4847924ddbd54727fa7b) to be on `validate_record` at the pids manager level in `manager.py` as a public method. My bad!

Here is what I understand you are suggesting:

Merge `_validate_record` into `_validate_pids` at the pids manager level and merge `validate_record` add `validate` at the provider level.

I think this is problematic. Let me try to break up my argument in 3 pieces so it is very clear to any reader:

#### 1- There is value in putting different things in different places

`validate` (I actually renamed it to `validate_pid` to address confusion problem) and `validate_record` on the provider do 2 different things. One of those things is to validate a **given** PID (an identifier/provider/client) and the *other* is to validate a record (actually the metadata) that potentially doesn't have a PID (e.g. when creating a new draft and letting InvenioRDM mint the DOI). Indeed, provider validation is also done on draft creation/save not just on publish. Passing the record for `validate_pid` is fine because we need to be able to tell if the pid is a duplicate.

For the confusion problem: if we walk through 2 scenarios in the shoes of a developer, things look like the following:

Scenario 1: Add a new provider

As a developer, I want to mint a new kind of PID in my instance. I therefore create a new `MyProvider` class that inherits from `PIDProvider`. Among the questions I need to answer are: what is a valid `identifier` (and potentially `provider`/`client`) for my PID? and what caracteristics must a record have in order to get minted my new kind of PID? Lo and behold two separate methods then exist to answer these 2 separate questions: `validate_pid` answers the first and `validate_record` answers the second.

Scenario 2: Maintain the codebase

I am a developer (of any experience level) and I need to fix a bug in how a Provider validates their identifier. `validate_pid` is the only method I need to know about and fix. I don't need to also understand the code related to what is a valid record according to this provider. You would need to understand that unrelated code if it were in the same method.

---

#### 2- The problem of very complex logic around "emptiness"

Conflating `validate_pid` and `validate_record` would also mean conflating the 2 loops that go through providers. One of those loops was only looking at given pids and the other was looking at given pids UNION required pids. In the case of given pids, an empty value is typically bad, while in the case of required pids an empty value is normal because the pid hasn't been minted yet (remember that validation happens at draft creation too). If both are funnelled in the same method we need to distinguish the cases where the PID (`identifier`/`provider`) is empty but shouldn't be, from the ones where it's empty but it's fine. That is incredibly error-prone. 

Pids are already very complex: we have to keep the draft/publish scenarios in mind, the pids that can change and those that can't, the external ones and the internal ones... Keeping the methods separate saves us from some complexity.

---

#### 3- The time and unintended consequences of extensive code changes

This goes back to the previous discussion we had. Conflating in one method means changing the interface of the errors outputted by the provider from a list of strings to a list of dicts. This involves changing a number of files that would not have been touched otherwise. This takes more time and opens up the risk of introducing other bugs.

---

To help move this along/identify what I am missing, I'd be happy to 

- have someone else's opinion on these points  
- take a look at actual alternative code suggestions
- have a video discussion about it in January. I am not sure I can be more precise than I am right now though :laughing: 

I also hope that it is clear that my intent is to better the codebase and the project, not to save myself from necessary work or for any other reason.

See you in 2023!

Reply:
Given that the returned errors are now the same type as before, there is no need to break the interface and rename `validate` to `validate_pid`.
The interface was very clear: `validate(pid, record)`. The developer will decide **internally** if it makes sense to split in that provider's implementation.

Example:
```
def validate(pid, record):
    _validate_pid(pid)
    _validate_record(record)
```

There is no point to over-engineer this, given that we don't know upfront if all provider's implementation will need to validate both pid and record metadata.
If there was no change in the interface, we could probably go with your solution, but now I don't see the advantage.

Finally, about the for-loops, I agree it is very complex, but let's answer in the comment above: it is not clear at all to me why we are checking for the required_schemes and why we do that in the `validate_record` method.

Reply:
I agree with @ntarocco.
In my opinion, interface of `validate(pid, record)` is flexible enough to accommodate any needed validations inside. In addition, the interface is clean and easily understandable - it does not introduce the separation for pid and record, which is valuable in cases where it is simply not needed, providing a more generic solution - exactly what you would expect from a framework - and leaves it to the developer to decide about validating their specific use case.

The required schemes check is also not clear to me.

Reply:
Ok... 3rd time's the charm :slightly_smiling_face: https://github.com/inveniosoftware/invenio-rdm-records/pull/1146/files ?  I *think* that new PR combines `validate_pid` and `validate_record` into `validate()` as you want (right?). Perhaps it makes clearer some of the consequences of that too.